File: BinarizationAndDenoising/GenerateNoise/GenerateNoise.py
import cv2 as cv
import numpy as np


def add_salt_pepper_noise(src):
    """
    生成椒盐噪点图像
    :param src:
    :return:
    """
    height, width = src.shape[:2]
    noise_num = 1000
    # 随机生成噪点坐标
    xs = np.random.randint(0, width, noise_num, dtype=np.int)
    ys = np.random.randint(0, height, noise_num, dtype=np.int)

    for i in range(noise_num):
        # 保证黑色和白色噪点间隔出现
        if i % 2 == 1:
            src[ys[i], xs[i]] = (255, 255, 255)
        else:
            src[ys[i], xs[i]] = (0, 0, 0)
    return src


def add_gaussian_noise(src):
    """
    生成高斯噪声图像
    :param src:
    :return:
    """
    noise = np.zeros(src.shape, src.dtype)
    # 各通道均值
    mean = (15, 15, 15)
    # 各通道方差
    stddev = (30, 30, 30)
    cv.randn(noise, mean, stddev)
    cv.imshow("gaussian noise", noise)
    dst = cv.add(src, noise)
    cv.imshow("gussian dst", dst)
    return dst


src = cv.imread("D:/Images/lena.jpg")
add_gaussian_noise(src)

height, width = src.shape[:2]
copy = np.copy(src)
copy = add_salt_pepper_noise(copy)
# 将原图和处理后的图像并排显示在一张图像上
result = np.zeros([height, 2 * width, 3], dtype=src.dtype)

result[0: height, 0: width, :] = src
result[0: height, width: width * 2, :] = copy
cv.putText(result, "src", (10, 30), cv.FONT_HERSHEY_COMPLEX, 2.0, (0, 255, 255), 1)
cv.putText(result, "salt pepper", (10 + width, 30), cv.FONT_HERSHEY_SIMPLEX, 2.0, (0, 255, 255), 1)
cv.imshow("salt pepper img", result)

cv.waitKey(0)
cv.destroyAllWindows()

File: BinarizationAndDenoising/GenerateNoise/GenerateNoise.cpp
#include <opencv2\opencv.hpp>
#include <iostream>

using namespace std;
using namespace cv;

//产生椒盐噪声
void add_salt_pepper_noise(Mat &src);
//产生高斯噪声
void add_gaussian_noise(Mat &src);

int main(int argc, char** argv)
{
	Mat src = imread("D:/Images/lena.jpg");
	if (src.empty())
	{
		printf("can not find the img");
		return -1;
	}
	namedWindow("src", CV_WINDOW_AUTOSIZE);
	imshow("src", src);
	add_salt_pepper_noise(src);
	add_gaussian_noise(src);
	waitKey(0);
	return 0;
}

void add_salt_pepper_noise(Mat &src)
{
	RNG random(10000);
	int width = src.cols;
	int height = src.rows;
	//噪声总数为10000
	int noise_nums = 1000;
	for (int i = 0; i < noise_nums; i++)
	{
		//均匀分布生成噪点的坐标   
		int x = random.uniform(0, height);
		int y = random.uniform(0, width);
		if (i % 2 == 1)
		{
			src.at<Vec3b>(y, x) = Vec3b(255, 255, 255);
		}
		else
		{
			src.at<Vec3b>(y, x) = Vec3b(0, 0, 0);
		}
	}
	imshow("salt pepper", src);
}

void add_gaussian_noise(Mat & src)
{
	Mat noise = Mat::zeros(src.size(), src.type());
	//生成高斯噪声图像, 第二个参数对应每个通道上均值，第三个参数对应每个通道上的方差
	randn(noise, (15, 15, 15), (30, 30, 30));
	Mat dst;
	//将噪点图像与原图像进行叠加
	add(src, noise, dst);
	imshow("gaussian noise", noise);
	imshow("gaussian dst", dst);
}




